06_gpu_and_ml/stable_diffusion/a1111_tunnel.py

```diff
@@ -0,0 +1,139 @@
+# ---
+# lambda-test: false
+# ---
+
+import os
+import socket
+import subprocess
+import time
+
+from modal import Dict, Image, Secret, Stub, asgi_app, forward
+
+stub = Stub("example-a1111")
+stub.urls = Dict.new()  # TODO: Hack because spawn() doesn't support generators.
+
+
+def wait_for_port(port: int):
+    while True:
+        try:
+            with socket.create_connection(("127.0.0.1", port), timeout=5.0):
+                break
+        except OSError:
+            time.sleep(0.1)
+
+
+@stub.function(
+    image=Image.debian_slim()
+    .apt_install(
+        "wget",
+        "git",
+        "libgl1",
+        "libglib2.0-0",
+        "google-perftools",  # For tcmalloc
+    )
+    .env({"LD_PRELOAD": "/usr/lib/x86_64-linux-gnu/libtcmalloc.so.4"})
+    .run_commands(
+        "git clone --depth 1 --branch v1.6.0 https://github.com/AUTOMATIC1111/stable-diffusion-webui /webui",
+        "python -m venv /webui/venv",
+        "cd /webui && . venv/bin/activate && "
+        + "python -c 'from modules import launch_utils; launch_utils.prepare_environment()' --xformers",
```

Comment:
could probably use `run_function` for this?

Reply:
I think the run_function would have to enter the venv again though :/